integration/ledger/snapshot_test.go

```diff
@@ -0,0 +1,1049 @@
+	Expect(err).NotTo(HaveOccurred())
+	Eventually(sess, n.EventuallyTimeout).Should(gexec.Exit(0))
+	orgGroup := &ordererext.DynamicOrdererOrgGroup{ConfigGroup: &common.ConfigGroup{}}
+	err = protolator.DeepUnmarshalJSON(bytes.NewBuffer(sess.Out.Contents()), orgGroup)
+	Expect(err).NotTo(HaveOccurred())
+	updatedConfig.ChannelGroup.Groups["Application"].Groups[org.Name] = orgGroup.ConfigGroup
+	nwo.UpdateConfig(n, orderer, channelID, currentConfig, updatedConfig, true, existingPeers[0], n.Peers...)
+}
+
+func joinBySnapshot(n *nwo.Network, orderer *nwo.Orderer, peer *nwo.Peer, channelID string, snapshotDir string, lastBlockInSnapshot int) {
+	channelHeight := nwo.GetMaxLedgerHeight(n, channelID, n.PeersWithChannel(channelID)...)
+
+	By(fmt.Sprintf("joining a peer via snapshot %s", snapshotDir))
+	n.JoinChannelBySnapshot(snapshotDir, orderer, peer)
+
+	By("calling JoinBySnapshotStatus")
+	checkStatus := func() []byte { return n.JoinBySnapshotStatus(orderer, peer) }
+	Eventually(checkStatus, n.EventuallyTimeout, 10*time.Second).Should(ContainSubstring("No joinbysnapshot operation is in progress"))
+
+	By("verifying blockchain info on peer " + peer.ID())
+	sess, err := n.PeerUserSession(peer, "Admin", commands.ChannelInfo{
+		ChannelID: channelID,
+	})
+	Expect(err).NotTo(HaveOccurred())
+	Eventually(sess, n.EventuallyTimeout).Should(gexec.Exit(0))
+	channelInfoStr := strings.TrimPrefix(string(sess.Buffer().Contents()[:]), "Blockchain info:")
+	var bcInfo = common.BlockchainInfo{}
+	err = json.Unmarshal([]byte(channelInfoStr), &bcInfo)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(bcInfo.Height).To(Equal(uint64(channelHeight)))
+	Expect(bcInfo.BootstrappingSnapshotInfo.LastBlockInSnapshot).To(Equal(uint64(lastBlockInSnapshot)))
+
+	By("waiting for all peers to have the same ledger height")
+	channelPeers := n.PeersWithChannel(channelID)
+	maxLedgerHeight := nwo.GetMaxLedgerHeight(n, channelID, channelPeers...)
+	nwo.WaitUntilEqualLedgerHeight(n, channelID, maxLedgerHeight, channelPeers...)
+}
+
+// verify the following functions for the peer after joining a channel by snapshot
+// - invoke/query chaincode
+// - upgrade chaincode
+// - channel config update
+// - join new channel
+// - qscc
+// - query history
+func verifyPeerAfterJoinBySnapshot(helper *marblesTestHelper, peer *nwo.Peer, channelID string, chaincode *nwo.Chaincode, newCCSequence string,
+	lastBlockInSnapshot int, oldTxid string, testKey string, availableHistoryResult []*marbleHistoryResult) []*marbleHistoryResult {
+
+	n := helper.Network
+	orderer := helper.orderer
+	peerID := peer.ID()
+
+	helper.invokeMarblesChaincode(chaincode.Name, peer, "transferMarble", testKey, "newowner_"+newCCSequence)
+
+	if newCCSequence != "" && newCCSequence != chaincode.Sequence {
+		chaincode.Version = newCCSequence + ".0"
+		chaincode.Sequence = newCCSequence
+		By("upgrading chaincode to version " + chaincode.Version + " on peer " + peerID)
+		nwo.DeployChaincode(n, channelID, orderer, *chaincode)
+	}
```

Comment:
This kind of side-effect in the helper function causes confusion in the main flow. I guess that it's cleaner to pass such details from the main flow.